app/scripts/controllers/permissions/index.js

```diff
@@ -0,0 +1,418 @@
+
+    if (ethAccounts) {
+
+      await this.validateExposedAccounts(accounts)
+
+      if (!ethAccounts.caveats) {
+        ethAccounts.caveats = []
+      }
+
+      // caveat names are unique, and we will only construct this caveat here
+      ethAccounts.caveats = ethAccounts.caveats.filter(c => (
+        c.name !== CAVEAT_NAMES.exposedAccounts
+      ))
+
+      ethAccounts.caveats.push(
+        {
+          type: 'filterResponse',
+          value: accounts,
+          name: CAVEAT_NAMES.exposedAccounts,
+        },
+      )
+    }
+  }
+
+  /**
+   * Validate an array of accounts representing accounts to be exposed
+   * to a domain. Throws error if validation fails.
+   *
+   * @param {string[]} accounts - An array of addresses.
+   */
+  async validateExposedAccounts (accounts) {
+
+    if (!Array.isArray(accounts) || accounts.length === 0) {
+      throw new Error('Must provide non-empty array of account(s).')
+    }
+
+    // assert accounts exist
+    const allAccounts = await this.keyringController.getAccounts()
+    accounts.forEach(acc => {
+      if (!allAccounts.includes(acc)) {
+        throw new Error(`Unknown account: ${acc}`)
+      }
+    })
+  }
+
+  /**
+   * Removes the given permissions for the given domain.
+   * @param {object} domains { origin: [permissions] }
+   */
+  removePermissionsFor (domains) {
+
+    Object.entries(domains).forEach(([origin, perms]) => {
+
+      this.permissions.removePermissionsFor(
+        origin,
+        perms.map(methodName => {
+
+          if (methodName === 'eth_accounts') {
+            this.notifyDomain(
+              origin,
+              { method: ACCOUNTS_CHANGED_NOTIFICATION, result: [] }
+            )
+          }
+
+          return { parentCapability: methodName }
+        })
+      )
+    })
+  }
+
+  /**
+   * Gets all caveats for the given origin and permission, or returns null
+   * if none exist.
+   *
+   * @param {string} permission - The name of the target permission.
+   * @param {string} origin - The origin that has the permission.
+   */
+  getCaveatsFor (permission, origin) {
+    return this.permissions.getCaveats(origin, permission) || null
+  }
+
+  /**
+   * Gets the caveat with the given name for the given permission of the
+   * given origin.
+   *
+   * @param {string} permission - The name of the target permission.
+   * @param {string} origin - The origin that has the permission.
+   * @param {string} caveatName - The name of the caveat to retrieve.
+   */
+  getCaveat (permission, origin, caveatName) {
+    return this.permissions.getCaveat(origin, permission, caveatName)
+  }
+
+  /**
+   * Removes all known domains and their related permissions.
+   */
+  clearPermissions () {
+    this.permissions.clearDomains()
+    this.notifyAllDomains({
+      method: ACCOUNTS_CHANGED_NOTIFICATION,
+      result: [],
+    })
+  }
+
+  /**
+   * Clears the permissions log.
+   */
+  clearLog () {
```

Comment:
These two methods appear to be unused. Does this mean the history and log will grow unbounded? :thinking:

Reply:
Not totally unbounded it looks like. The log store has a limit of 100, and the history only stores a few pieces of data per origin.